Update 'Compiling with Mono': iOS, WASM, AOT and BCL (#3265)
godotengine/godot-docs: development/compiling/compiling_with_mono.rst

The ``Tools`` subdirectory contains tools required by the editor, like the
``GodotTools`` assemblies and its dependencies.

Building the Mono runtime
-------------------------

When building Godot for the desktop, you will likely use the pre-built Mono runtime
that is installed on your system. This likely won't be the case when targeting other

> platforms like Android, iOS and WebAssembly. For those platforms you will have to

Reviewer: ```suggestion
platforms like Android, iOS and WebAssembly. You will have to build the Mono runtime
```

> build the Mono runtime yourself.

Reviewer: ```suggestion
yourself for those platforms.
```

::

    mkdir -p bin/ios
    mkdir -p bin/ios/iphone-mono-libs

    lipo -create bin/libmonosgen-2.0.iphone.arm64.a bin/libmonosgen-2.0.iphone.x86_64.a -output bin/ios/iphone-mono-libs/libmonosgen-2.0.iphone.fat.a
    lipo -create bin/libmono-native.iphone.arm64.a bin/libmono-native.iphone.x86_64.a -output bin/ios/iphone-mono-libs/libmono-native.iphone.fat.a
    lipo -create bin/libmono-profiler-log.iphone.arm64.a bin/libmono-profiler-log.iphone.x86_64.a -output bin/ios/iphone-mono-libs/libmono-profiler-log.iphone.fat.a

    # The Mono libraries for the interpreter are not available for simulator builds
    lipo -create bin/libmono-ee-interp.iphone.arm64.a -output bin/ios/iphone-mono-libs/libmono-ee-interp.iphone.fat.a
    lipo -create bin/libmono-icall-table.iphone.arm64.a -output bin/ios/iphone-mono-libs/libmono-icall-table.iphone.fat.a
    lipo -create bin/libmono-ilgen.iphone.arm64.a -output bin/ios/iphone-mono-libs/libmono-ilgen.iphone.fat.a

The ``iphone-mono-libs`` folder must be distributed with the export templates.
The Godot editor will look for the libraries in ``<templates>/iphone-mono-libs/lib<name>.iphone.fat.a``.

Targeting WebAssembly
---------------------

> As of now, building for WebAssembly is the same both when the mono module is enabled and when it is disabled.

Reviewer: ```suggestion
Building for WebAssembly currently involves the same process regardless of whether the Mono module is enabled.
```

Alternatively, Godot will look for them in the following locations:

+-------------------+---------------------------------+
|      Android      |  ``<templates>/bcl/monodroid``  |
+-------------------+---------------------------------+
|        iOS        |  ``<templates>/bcl/monotouch``  |
+-------------------+---------------------------------+
|    WebAssembly    |    ``<templates>/bcl/wasm``     |
+-------------------+---------------------------------+
|  Linux and macOS  |   ``<templates>/bcl/net_4_x``   |
+-------------------+---------------------------------+
|      Windows      | ``<templates>/bcl/net_4_x_win`` |
+-------------------+---------------------------------+

As of now, we're assuming the same BCL profile can be used for both Linux and macOS,
but this may change in the future as they're not guaranteed to be the same
(as is the case with the Windows BCL).

If the target platform is the same as the platform of the Godot editor,
then the editor will use the BCL it's running on (``<data_folder>/Mono/lib/mono/4.5``)
if it cannot find the BCL in the export templates.

AOT cross-compilers
-------------------

> In order to do Ahead Of Time (AOT) compilation for other platforms, Godot needs to have

Reviewer: ```suggestion
To perform ahead-of-time (AOT) compilation for other platforms, Godot needs to have
```
